process_hierarchy.py

```diff
@@ -0,0 +1,478 @@
+    datafields and not assiging it a key (which is what I've wound up
+    doing)
+    """
+    m = hashlib.sha256()
+    m.update((f"{str(high_key)}+{str(mid_key)}+{str(low_key)}").encode("utf-8"))
+    return m.hexdigest()
+
+
+def create_hierarchy_files(hierarchy):
+    """
+    This builds the hierarchy files from the data coming from coldfront, keycloak
+    and historic data from the database
+    """
+    # construct hierarchy.csv
+    with open("hierarchy.csv", "w", encoding="utf-8") as hierarchy_file:
+        for rec_id, rec in hierarchy["institution"].items():
+            unique_key = create_xdmod_key(rec_id)
+            hierarchy_file.write(f"{unique_key},{rec['name']},\n")
+        for l2_id, l2 in hierarchy["field-of-science"].items():
+            unique_key = create_xdmod_key(
+                hierarchy["inistitution"][l2["parent_id"]], l2_id
+            )
+            parent_key = create_xdmod_key(l2["parent_id"])
+            hierarchy_file.write(f"{unique_key},{l2['name']},{parent_key}\n")
+        for l3_id, l3_rec in hierarchy["pi"].items():
+            unique_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+                l3_id,
+            )
+            parent_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+            )
+            hierarchy_file.write(f"{unique_key},{l3_rec['name']},{parent_key}\n")
+
+    # constrcut groups.csv (level 4 of the hierarchy, though this is a mapping table between pi/group and the hiearachy)
+    with open("group.csv", "w", encoding="utf-8") as group_file:
+        for l4_id, l4_rec in hierarchy["project"].items():
+            l3_rec = hierarchy["pi"][l4_rec["parent_id"]]
+            unique_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+                l3_rec["id"],
+            )
+            group_file.write(f'"{l4_rec["name"]}", "{unique_key}" \n')
+
+    # constrcut pi2project
+    with open("pi2project.csv", "w", encoding="utf-8") as pi2project_file:
+        for l5 in hierarchy["cloud-project"].values():
+            l4_rec = hierarchy["cloud-project"][l5["parent_id"]]
+            l4_id = l4_rec["id"]
+            pi2project_file.write(
+                f"{hierarchy['cloud-project'][l4_id]['name']}, {l5['name']}\n"
+            )
+
+    # construct the names (rename projects)
+
+
+def find_hierarchy_id(name, dictionary):
+    """returns the hierarchy_id if found in the dictionary"""
+    for rec in dictionary.values():
+        if name == rec["name"]:
+            return rec["id"]
+    return None
+
+
+def get_hierarchy_from_db(cursor):
+    """This fetches the data from the database and only keeps the most recent record based on its id"""
```

Comment:
I will update the comment as it is not precise enough.

Reply:
The way around this is to have a more complicated database design.  One common way to accomplish this is to keep the most recent record in current table and move records from the current table into a history table.  For example, this worked well for a database that I worked with 10K - 100k elements.  With the current table, you never have to use a limit by one, you just dump the entire table and you get the table structure.  I felt that this was a little too much for a MVP as it would take too long to implement.  This is fast for the current dataset and slower for history.

Basically, what I have implemented so far is just the history table, as I see that as an MVP.  Going forward, without modifying this part, we could add a table containing the id and most recent timestamp for that ID.  With this, a single join with no need for a limit by 1 to pull the most recent record from the history table, and probably the easiest to implement with mariadb.

With a database, like oracle, you could have the current table with the same table structure as the history table.  Then add a store procedure that gets a record from the current table and insert that record into the history table upon an update to the current table.  That way, your programs could update the current table, and the history is tracked.